standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java

```diff
@@ -0,0 +1,311 @@
+    if (this.password == null) {
+      String keystore = connector.getParameters().get(JDBC_KEYSTORE);
+      String key = connector.getParameters().get(JDBC_KEY);
+      try {
+        char[] keyValue = MetastoreConf.getValueFromKeystore(keystore, key);
+        if (keyValue != null)
+          this.password = new String(keyValue);
+      } catch (IOException i) {
+        LOG.warn("Could not read key value from keystore");
+      }
+    }
+
+    try {
+      warehouse = new Warehouse(MetastoreConf.newMetastoreConf());
+    } catch (MetaException e) { /* ignore */ }
+  }
+
+  @Override public void open() throws ConnectException {
+    try {
+      Class.forName(driverClassName);
+      handle = DriverManager.getConnection(jdbcUrl, username, password);
+      isOpen = true;
+    } catch (ClassNotFoundException cnfe) {
+      LOG.warn("Driver class not found in classpath:" + driverClassName);
+      throw new RuntimeException("Driver class not found:" + driverClassName);
+    } catch (SQLException sqle) {
+      LOG.warn("Could not connect to remote data source at " + jdbcUrl);
+      throw new ConnectException("Could not connect to remote datasource at " + jdbcUrl + ",cause:" + sqle.getMessage());
+    }
+  }
+
+  protected Connection getConnection() {
+    try {
+      if (!isOpen)
+        open();
+    } catch (ConnectException ce) {
+      throw new RuntimeException(ce.getMessage());
+    }
+
+    if (handle instanceof Connection)
+      return (Connection)handle;
+
+    throw new RuntimeException("unexpected type for connection handle");
+  }
+
+  @Override public void close() {
+    if (isOpen) {
+      try {
+        ((Connection)handle).close();
+      } catch (SQLException sqle) {
+        LOG.warn("Could not close jdbc connection to " + jdbcUrl, sqle);
+      }
+    }
+  }
+
+  /**
+   * Returns Hive Table objects from the remote database for tables that match a name pattern.
+   * @return List A collection of objects that match the name pattern, null otherwise.
+   * @throws MetaException To indicate any failures with executing this API
+   * @param regex
+   */
+  @Override public abstract List<Table> getTables(String regex) throws MetaException;
+
+  /**
+   * Returns a list of all table names from the remote database.
+   * @return List A collection of all the table names, null if there are no tables.
+   * @throws MetaException To indicate any failures with executing this API
+   */
+  @Override public List<String> getTableNames() throws MetaException {
+    ResultSet rs = null;
+    try {
+      rs = getConnection().getMetaData().getTables(scoped_db, null, null, new String[] { "TABLE" });
+      if (rs != null) {
+        List<String> tables = new ArrayList<String>();
+        while(rs.next()) {
+          tables.add(rs.getString(3));
+        }
+        return tables;
+      }
+    } catch (SQLException sqle) {
+      LOG.warn("Could not retrieve table names from remote datasource, cause:" + sqle.getMessage());
```

Comment:
Don't we want to throw the SQLException forward here? We can wrap it in a MetaException and let the higher layers handle it?

Reply:
Good catch, fixed.